File: src/main/java/com/jackharrhy/storefront/Storage.java
package com.jackharrhy.storefront;

import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.entity.Player;
import org.jdbi.v3.core.Jdbi;

import java.io.File;
import java.time.Instant;
import java.util.List;
import java.util.Optional;
import java.util.logging.Level;
import java.util.logging.Logger;
import java.util.stream.Collectors;

public class Storage {
    private Logger logger;
    private Jdbi jdbi;

    private static String serializeLocation(Location location) {
        return "" + location.getWorld().getName()
                + ":" + location.getX()
                + ":" + location.getY()
                + ":" + location.getZ();
    }

    private Location deserializeLocation(String serializedLocation) {
        String[] splitLocation = serializedLocation.split(":");
        return new Location(
                Bukkit.getWorld(splitLocation[0]),
                Double.parseDouble(splitLocation[1]),
                Double.parseDouble(splitLocation[2]),
                Double.parseDouble(splitLocation[3])
        );
    }

    public Storage(Logger logger, String fileName) {
        this.logger = logger;

        File dbFile = new File(fileName);

        if (!dbFile.exists()) {
            try {
                dbFile.createNewFile();
            } catch (Exception e) {
                logger.log(Level.SEVERE, e.getMessage());
                e.printStackTrace();
            }
        }

        String connectionString = "jdbc:sqlite:" + fileName;
        this.jdbi = Jdbi.create(connectionString);
        this.initialize();
    }

    private void initialize() {
        String createChestTableSql = "CREATE TABLE IF NOT EXISTS chest ("
                + "	id INTEGER PRIMARY KEY,"
                + " owner TEXT NOT NULL, "
                + " location TEXT NOT NULL, "
                + "	contents TEXT NOT NULL, "
                + "	modified INTEGER NOT NULL"
                + ");";
        jdbi.useHandle(handle -> handle.execute(createChestTableSql));
    }

    public Boolean updateStorefront(Player owner, Location location, String contents) {
        String locationSerialized = serializeLocation(location);

        String insertOrUpdateStorefrontSql = "INSERT OR REPLACE INTO chest (id, owner, location, contents, modified) " +
                "VALUES ((SELECT id FROM chest WHERE location = :location), :owner, :location, :contents, :modified)";

        int updated = jdbi.withHandle(handle -> handle.createUpdate(insertOrUpdateStorefrontSql)
                .bind("location", locationSerialized)
                .bind("owner", owner.getUniqueId().toString())
                .bind("contents", contents)
                .bind("modified", Instant.now().getEpochSecond())
                .execute());

        if (updated == 1) {
            return true;
        } else if (updated == 0) {
            return false;
        } else {
            logger.log(Level.SEVERE, "Updated more than one storefront on a single call");
            return null;
        }
    }

    public Boolean updateStorefront(Location location, String contents) {
        String locationSerialized = serializeLocation(location);

        String updateStorefrontSql = "UPDATE chest SET contents = :contents, modified = :modified " +
                "WHERE id = (SELECT id FROM chest WHERE location = :location)";

        int updated = jdbi.withHandle(handle -> handle.createUpdate(updateStorefrontSql)
                .bind("location", locationSerialized)
                .bind("contents", contents)
                .bind("modified", Instant.now().getEpochSecond())
                .execute());

        if (updated == 1) {
            return true;
        } else if (updated == 0) {
            return false;
        } else {
            logger.log(Level.SEVERE, "Updated more than one storefront on a single call");
            return null;
        }
    }

    public Boolean removeStorefront(Player owner, Location location) {
        String removeStorefrontSql = "DELETE FROM chest WHERE location = ? AND owner = ?";

        int removed = jdbi.withHandle(handle -> handle.execute(
                removeStorefrontSql, serializeLocation(location), owner.getUniqueId().toString()
        ));

        if (removed == 1) {
            return true;
        } else if (removed == 0) {
            return false;
        } else {
            logger.log(Level.SEVERE, "Removed more than one storefront on a single call");
            return null;
        }
    }

    public Boolean removeStorefront(Location location) {
        String removeStorefrontSql = "DELETE FROM chest WHERE location = ?";

        int removed = jdbi.withHandle(handle -> handle.execute(
                removeStorefrontSql, serializeLocation(location)
        ));

        if (removed == 1) {
            return true;
        } else if (removed == 0) {
            return false;
        } else {
            logger.log(Level.SEVERE, "Removed more than one storefront on a single call");
            return null;
        }
    }

    public List<String> getAllContents() {
        return jdbi.withHandle(handle -> handle.createQuery("SELECT contents FROM chest")
                .mapTo(String.class)
                .list());
    }

    public List<Location> getAllLocations() {
        return jdbi.withHandle(handle -> handle.createQuery("SELECT location FROM chest")
                .mapTo(String.class)
                .list()
                .stream()
                .map((string) -> deserializeLocation(string))
                .collect(Collectors.toList()));
    }

    public Optional<String> owner(Location location) {
        String getOwnerFromChestSql = "SELECT owner FROM chest WHERE location = ?";
        return jdbi.withHandle(handle -> handle.select(getOwnerFromChestSql, serializeLocation(location))
                .mapTo(String.class)
                .findFirst());
    }
}

File: src/main/kotlin/com/jackharrhy/storefront/Storage.kt
package com.jackharrhy.storefront

import com.google.gson.Gson
import com.google.gson.JsonElement
import com.google.gson.JsonParser
import org.jdbi.v3.core.Jdbi
import org.sqlite.SQLiteDataSource
import java.sql.ResultSet
import java.util.concurrent.atomic.AtomicLong

data class RefreshTarget(val id: Int, val location: String, val contents: String, val modified: Long)
data class RefreshChange(val target: RefreshTarget, val contents: String?)
data class Owner(val uuid: String, val name: String)
data class Content(val id: Int, val owner: JsonElement, val contents: JsonElement, val description: JsonElement)

class Storage(fileName: String) {
    private val gson = Gson()
    private val modifiedClock = AtomicLong(System.currentTimeMillis())
    private val jdbi = Jdbi.create(SQLiteDataSource().apply {
        url = "jdbc:sqlite:$fileName"
        config.setBusyTimeout(5000)
    })

    init {
        jdbi.useHandle<RuntimeException> { handle ->
            handle.execute("PRAGMA journal_mode=WAL")
            handle.execute("""
                CREATE TABLE IF NOT EXISTS chest (
                    id INTEGER PRIMARY KEY,
                    owner TEXT NOT NULL,
                    location TEXT NOT NULL,
                    contents TEXT NOT NULL,
                    modified INTEGER NOT NULL,
                    description TEXT NOT NULL
                )
            """.trimIndent())
            handle.execute("CREATE INDEX IF NOT EXISTS chest_location ON chest(location)")
            val latest = handle.createQuery("SELECT COALESCE(MAX(modified), 0) FROM chest")
                .mapTo(Long::class.java).one()
            modifiedClock.updateAndGet { maxOf(it, latest) }
        }
    }

    private fun nextModified(): Long = modifiedClock.updateAndGet { maxOf(it + 1, System.currentTimeMillis()) }

    private fun readContent(row: ResultSet) = Content(
        row.getInt("id"),
        JsonParser.parseString(row.getString("owner")),
        JsonParser.parseString(row.getString("contents")),
        JsonParser.parseString(row.getString("description"))
    )

    val allContents: List<Content>
        get() = jdbi.withHandle<List<Content>, RuntimeException> { handle ->
            handle.createQuery("SELECT id, owner, contents, description FROM chest")
                .map { row, _ -> readContent(row) }.list()
        }

    fun storefrontContentsById(id: Int): Content? = jdbi.withHandle<Content?, RuntimeException> { handle ->
        handle.select("SELECT id, owner, contents, description FROM chest WHERE id = ?", id)
            .map { row, _ -> readContent(row) }.findOne().orElse(null)
    }

    fun storefrontLocationString(id: Int): String? = jdbi.withHandle<String?, RuntimeException> { handle ->
        handle.select("SELECT location FROM chest WHERE id = ?", id)
            .mapTo(String::class.java).findOne().orElse(null)
    }

    fun ownerUUID(location: String): String? = jdbi.withHandle<String?, RuntimeException> { handle ->
        handle.select("SELECT json_extract(owner, '$.uuid') FROM chest WHERE location = ?", location)
            .mapTo(String::class.java).findFirst().orElse(null)
    }

    fun newStorefront(owner: Owner, location: String, contents: String, description: Array<String>): Boolean =
        jdbi.withHandle<Boolean, RuntimeException> { handle ->
            handle.createUpdate("""
                INSERT OR REPLACE INTO chest (id, owner, location, contents, modified, description)
                VALUES ((SELECT id FROM chest WHERE location = :location),
                    :owner, :location, :contents, :modified, :description)
            """.trimIndent())
                .bind("owner", gson.toJson(owner))
                .bind("location", location)
                .bind("contents", contents)
                .bind("modified", nextModified())
                .bind("description", gson.toJson(description))
                .execute() == 1
        }

    fun updateStorefront(location: String, contents: String, description: Array<String>): Boolean =
        jdbi.withHandle<Boolean, RuntimeException> { handle ->
            handle.createUpdate("""
                UPDATE chest SET contents = :contents, modified = :modified, description = :description
                WHERE id = (SELECT id FROM chest WHERE location = :location)
            """.trimIndent())
                .bind("location", location)
                .bind("contents", contents)
                .bind("modified", nextModified())
                .bind("description", gson.toJson(description))
                .execute() == 1
        }

    fun removeStorefront(ownerUuid: String, location: String): Boolean =
        jdbi.withHandle<Boolean, RuntimeException> { handle ->
            handle.execute("DELETE FROM chest WHERE location = ? AND json_extract(owner, '$.uuid') = ?", location, ownerUuid) == 1
        }

    fun refreshTargets(): List<RefreshTarget> = jdbi.withHandle<List<RefreshTarget>, RuntimeException> { handle ->
        handle.createQuery("SELECT id, location, contents, modified FROM chest ORDER BY id")
            .map { row, _ -> RefreshTarget(row.getInt("id"), row.getString("location"), row.getString("contents"), row.getLong("modified")) }
            .list()
    }

    fun applyRefresh(changes: List<RefreshChange>): Int {
        if (changes.isEmpty()) return 0
        return jdbi.inTransaction<Int, RuntimeException> { handle ->
            changes.sumOf { (target, contents) ->
                val statement = if (contents == null) {
                    handle.createUpdate("DELETE FROM chest WHERE id = :id AND modified = :previous")
                } else {
                    handle.createUpdate("UPDATE chest SET contents = :contents, modified = :modified WHERE id = :id AND modified = :previous")
                        .bind("contents", contents).bind("modified", nextModified())
                }
                statement.bind("id", target.id).bind("previous", target.modified).execute()
            }
        }
    }
}
